src/treebuilder.rs

```diff
@@ -0,0 +1,223 @@
+        let filename = try!(filename.as_ref().into_c_string());
+        unsafe {
+            let ret = raw::git_treebuilder_get(self.raw, ::call::convert(&filename));
+            Ok(TreeEntry::from_raw_const(ret))
+        }
+    }
+
+    /// Add or update an entry in the builder
+    ///
+    /// No attempt is made to ensure that the provided Oid points to
+    /// an object of a reasonable type (or any object at all)
+    pub fn insert<P: AsRef<OsStr>>(&mut self, filename: P, oid: Oid,
+                                   filemode: i32) -> Result<TreeEntry, Error> {
+        use raw::FromPrimitive;
+        let filename = try!(filename.as_ref().into_c_string());
+        let filemode = raw::git_filemode_t::from_i32(filemode)
+            .expect("Invalid filemode passed to TreeBuilder::insert");
+
+        let mut ret = 0 as *const raw::git_tree_entry;
+        unsafe {
+            try_call!(raw::git_treebuilder_insert(&mut ret, self.raw, filename,
+                                                  oid.raw(), filemode));
+            Ok(TreeEntry::from_raw_const(ret))
+        }
+    }
+
+    /// Remove an entry from the builder by its filename
+    pub fn remove<P: AsRef<OsStr>>(&mut self, filename: P) -> Result<(), Error> {
+        let filename = try!(filename.as_ref().into_c_string());
+        unsafe {
+            try_call!(raw::git_treebuilder_remove(self.raw, filename));
+        }
+        Ok(())
+    }
+
+    /// Selectively remove entries from the tree
+    pub fn filter<F>(&mut self, mut filter: F)
+                     where F: FnMut(&TreeEntry) -> FilterResult {
+        let mut cb: &mut FilterCb = &mut filter;
+        let ptr = &mut cb as *mut _;
+        unsafe {
+            raw::git_treebuilder_filter(self.raw, filter_cb, ptr as *mut _);
+            panic::check();
+        }
+    }
+
+    /// Write the contents of the TreeBuilder as a Tree object and
+    /// return its Oid
+    pub fn write(&self) -> Result<Oid, Error> {
+        let mut raw = raw::git_oid { id: [0; raw::GIT_OID_RAWSZ] };
+        unsafe {
+            try_call!(raw::git_treebuilder_write(&mut raw, self.raw()));
+            Ok(Binding::from_raw(&raw as *const _))
+        }
+    }
+}
+
+type FilterCb<'a> = FnMut(&TreeEntry) -> FilterResult + 'a;
+
+/// Action to be taken for an entry in TreeBuilder::filter
+// We use an enum here rather than a bool because libgit2 and libstd's
+// filtering functions disagree on the meaning of the return value, so
+// either choice would be confusing.
+pub enum FilterResult {
+    /// The TreeEntry should be kept in the builder
+    Keep,
+    /// The TreeEntry should be removed from the builder
+    Discard,
+}
+
+wrap_env! {
+    fn filter_cb(entry: *const raw::git_tree_entry,
+                 payload: *mut c_void) -> c_int {
+        unsafe {
+            // There's no way to return early from git_treebuilder_filter, so
+            // just keep panicing until it stops calling us.
+            panic::check();
```

Comment:
I think this may need a new helper method to return early because otherwise this'll just keep taking the same panic result out of TLS and putting it back in (with another box).